tools/indexscan/main.go

```diff
@@ -251,6 +252,96 @@ func (s CommitsNearPublishStrategy) Search(ctx context.Context, r *git.Repositor
 	return
 }
 
+// ExhaustiveTreeSearchStrategy
+type ExhaustiveTreeSearchStrategy struct {
+}
+
+// Search returns the set of matching commits along with the number of matches.
+func (s ExhaustiveTreeSearchStrategy) Search(ctx context.Context, r *git.Repository, hashes []string) (closest []string, matched, total int, err error) {
+	// Find subset of file hashes that are present in the repo.
+	files := make(map[plumbing.Hash]int)
+	var i int
+	for _, h := range hashes {
+		_, err = r.BlobObject(plumbing.NewHash(h))
+		if err == plumbing.ErrObjectNotFound {
+			// Skip files not present in repo.
+			err = nil
+		} else if err != nil {
+			return
+		} else {
+			files[plumbing.NewHash(h)] = i
+			i++
+		}
+	}
+	// Iterate over repo trees to create an ordering.
+	trees := make(map[plumbing.Hash]int)
+	i = 0
+	for ti, _ := r.TreeObjects(); ; {
+		t, err := ti.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			panic(err)
+		} else {
+			trees[t.Hash] = i
+			i++
+		}
+	}
+	total = len(files)
+	if total == 0 {
+		err = errors.New("repo contains no matching files")
+		return
+	}
+	// Recursively build set of matching files for each tree.
+	matches := bitmap.NewBatch(len(files), len(trees))
+	processed := bitmap.New(len(trees))
+	ti, _ := r.TreeObjects()
+	ti.ForEach(func(t *object.Tree) error {
+		processTree(t, files, trees, matches, processed)
+		return nil
+	})
+	// Search through all commits for the one whose tree has the most matches.
+	ci, _ := r.CommitObjects()
+	err = ci.ForEach(func(c *object.Commit) error {
+		if !processed.Get(trees[c.TreeHash]) {
+			return errors.New("unprocessed tree")
+		}
+		count := matches[trees[c.TreeHash]].Count()
+		if matched < count {
+			matched = count
+			closest = closest[:0]
+		}
+		if matched == count {
+			closest = append(closest, c.Hash.String())
+		}
+		return nil
+	})
+	return
+}
+
+// processTree recursively determines the presence of a set of files in the given git Tree and records them.
+func processTree(t *object.Tree, files, trees map[plumbing.Hash]int, matches []bitmap.Bitmap, processed *bitmap.Bitmap) {
+	if processed.Get(trees[t.Hash]) {
+		return
+	}
+	for _, e := range t.Entries {
+		switch e.Mode {
+		case filemode.Dir:
+			if !processed.Get(trees[t.Hash]) {
+				st, _ := t.Tree(e.Name)
+				processTree(st, files, trees, matches, processed)
+			}
+			matches[trees[t.Hash]].Or(&matches[trees[e.Hash]])
+		case filemode.Submodule, filemode.Symlink:
+			continue
+		default:
+			matches[trees[t.Hash]].Set(files[e.Hash])
```

Comment:
I think I must be missing something, why doesn't files[e.Hash] sometimes cause a key error? IIUC `files` only contains the files originally passed into Search as `hashes`, but e.Hash could be any hash of any file in the entire git history (presumably not in `hashes`).

Reply:
there's no lookup error it just maps to the default `int` value (i.e. 0) if it's not found: [link](https://go.dev/blog/maps#:~:text=If%20the%20requested%20key%20doesn%E2%80%99t%20exist%2C%20we%20get%20the%20value%20type%E2%80%99s%20zero%20value.%20In%20this%20case%20the%20value%20type%20is%20int%2C%20so%20the%20zero%20value%20is%200%3A)

Reply:
Does that mean any tree with files that don't exist in `hashes` is marked as matching `hashes[0]`?

Reply:
Discussed offline, the previous behavior was ~always~ counting a match for hashes[0], and when trying to find a "best match" that's not a very obvious bug. The latest version checks for the hash in `files` before counting it as a match.